exonum/src/blockchain/service.rs

```diff
@@ -63,6 +65,204 @@ pub trait Transaction: Message + ExonumJson + 'static {
     fn execute(&self, fork: &mut Fork);
 }
 
+/// `TransactionSet` trait describes a type which is an `enum` of several transactions.
+/// The implementation of this trait is generated automatically by the `transactions!`
+/// macro.
+pub trait TransactionSet
+    : Into<Box<Transaction>> + DeserializeOwned + Serialize + Clone {
```

Comment:
Is `Into<Box<Transaction>>` really needed? Seems like the code compiles fine without it.

Reply:
It is used here https://github.com/exonum/exonum/pull/457/files#diff-29918e4d0be4fa7170f31c0e3e9152f8R822 and is tested by doctest for `Service`.

Reply:
If the supertrait is removed (but not `impl Into<Box<Transaction>>` for types generated by `transactions!`; I'm not disputing that it's needed), all lib tests and the `Service` doc test still work fine (at least, for me locally). So, the question is: is the supertrait (not implementation!) needed, and what does it convey logically? More generally, I've got an impression (maybe, a mistaken one - would be glad to be corrected!) that it's considered a best practice in Rust not to use supertraits unless completely necessary.

Reply:
Hm, indeed 🤔 

The idea is that one can write a generic `tx_from_raw<S: TransactionSet>` and a generic `POST /transactions` methods using only `: TransactionSet` bound, and not `TransactionSet + Into`. This is not done in this PR to reduce its scope, but the approach should be similar to https://github.com/exonum/exonum/pull/424.

Given that we generate `TransactionSet`s via macros, and we always generate the `Into` impl, I don't think that the bound is particularly problematic. 

However I'll try to take a holistic look at the trait hierarchy  for `Transaction`, `Message` and `StorageValue` and see if something could be drastically simplified there!